src/test/java/com/google/sps/UserLoginTest.java

```diff
@@ -0,0 +1,107 @@
+package com.google.sps;
+
+import static org.mockito.Mockito.when;
+
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.sps.servlets.UserLoginServlet;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class UserLoginTest {
+
+  @Mock private HttpServletRequest request;
+
+  @Mock private HttpServletResponse response;
+
+  @Mock private UserLoginServlet userServlet;
+
+  private LocalServiceTestHelper helperLoggedIn =
+    new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+        .setEnvIsAdmin(true).setEnvIsLoggedIn(true);
+  
+  private LocalServiceTestHelper helperLoggedOut =
+    new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+        .setEnvIsAdmin(true).setEnvIsLoggedIn(false);
+  
+  private JsonObject jsonSetUp() throws ServletException, IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String response = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(response);
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
+    
+    return responseJsonObject;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test
+  public void loggedInUserReturnsLogOutUrl() throws ServletException, IOException  {
+
+    helperLoggedIn.setUp();
+    
+    JsonObject responseJsonObject = jsonSetUp();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
+
+    Assert.assertFalse(logOutUrl.isEmpty());
+    helperLoggedIn.tearDown();
+  }
+
+
+  @Test
+  public void LogOutUrlContainsLogOut() throws ServletException, IOException  {
+
+    helperLoggedIn.setUp();
+    
+    JsonObject responseJsonObject = jsonSetUp();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
+    
+    Assert.assertTrue(logOutUrl.contains("logout"));
```

Comment:
it seems like if logOutUrl.contains("logout") is true then logOutUrl.isEmpty() would be false by definition. Consider if it's necessary to have two separate tests for verifying the logout url contents.

What might make sense here is to have a separate test that validates that the login url is empty when the user is logged in and that the logout url is empty when the user is logged out.

Reply:
Is there a reason we need both assertFalse(logOutUrl.isEmpty()) and assertTrue(logOutUrl.contains("logout"))?